Source/Internal/IGListPreprocessor.m

```diff
@@ -0,0 +1,98 @@
+//
+//  IGListPreprocessor.m
+//  IGListKit
+//
+//  Created by Adlai Holler on 2/20/17.
+//  Copyright © 2017 Instagram. All rights reserved.
+//
+
+#import "IGListPreprocessor.h"
+
+#import "IGListAssert.h"
+#import "IGListPreprocessingContext.h"
+#import "IGListPreprocessingDelegate.h"
+#import "IGListSectionController.h"
+#import "IGListSectionMap.h"
+
+@implementation IGListPreprocessor {
+	NSMapTable<IGListPreprocessingContext *, id<IGListPreprocessingDelegate>> *_contextToDelegateMap;
+	IGListSectionMap *_sectionMap;
+	dispatch_group_t _group;
+	dispatch_block_t _completion;
+	BOOL _ranCompletion;
+}
+
+- (instancetype)initWithSectionMap:(IGListSectionMap *)sectionMap {
+	if (self = [super init]) {
+		_contextToDelegateMap = [NSMapTable strongToStrongObjectsMapTable];
+		_sectionMap = [sectionMap copy];
+		_group = dispatch_group_create();
+	}
+	return self;
+}
+
+- (void)schedulePreprocessingWithContainerSize:(CGSize)containerSize completion:(nonnull dispatch_block_t)completionBlock {
+	IGAssertMainThread();
+	_completion = completionBlock;
```

Comment:
++ assertion. Shouldn't ever be running 2 preprocessing jobs at a time:

- Preprocess, dispatching work to concurrent queues, wait til done
- Then `performBatchUpdates:`
- Rely on same IGLK infra to prevent multiple batch updates happening at the same time